### ISSUE TITLE
Session re-keying

### ISSUE THREAD

**Issue by aantron** [aantron/dream#13]
When a session key is half-expired, Dream refreshes its expiration time and updates the client cookie:

https://github.com/aantron/dream/blob/4b3ec4bf1d8481f163eafef367f62ade91ae6fbf/src/middleware/session.ml#L162-L165

Dream should [issue a new session key instead](https://cheatsheetseries.owasp.org/cheatsheets/Session_Management_Cheat_Sheet.html#renewal-timeout), but continue to accept the old key until it expires at its original expiration time.

**Comment by cemerick** [comment]
I was surprised to see Dream storing empty sessions. I thought it was probably just a low-priority TODO, but there's a "Lazy sessions" item here, with a stricken section that (I think?) implies that there's some value in persisting empty sessions? Can you clarify what's going on there?

**Comment by aantron** [comment]
I wasn't able to come to a decision in the summer about whether the existence of the session is itself important, even if the session is empty. So I chose for now that if a request passes through the session middleware, it always gets a session, even if nothing is stored in it. This allows, at least, recognizing subsequent requests as belonging to the same session. I'm very open to discussing whether this is right or not.

**Comment by cemerick** [comment]
Seems there are two (IMO) good reasons not to create sessions until they're necessary:

1. In practice, it will result in a (maybe heavy) slurry of junk `dream_session` records in the database / redis / whatever. Not great, not terrible (but also, why?)
2. More problematic is that always issuing session cookies means one can't (easily) make an app be cookie-free until login (a helpful attribute vis a vis GDPR compliance, etc).

**Comment by aantron** [comment]
> Seems there are two (IMO) good reasons not to create sessions until they're necessary:
> 
> 1. In practice, it will result in a (maybe heavy) slurry of junk `dream_session` records in the database / redis / whatever. Not great, not terrible (but also, why?)
> 2. More problematic is that always issuing session cookies means one can't (easily) make an app be cookie-free until login (a helpful attribute vis a vis GDPR compliance, etc).

Indeed, I completely agree, and now I recall that I was considering these issues as well. And the "Lazy sessions" bullet point is still open, to be addressed. That part of it is stricken is confusing. It is confusing even me at this point, as a reader of this issue months later :P But your comment serves as a good clarification, thank you!

> probably just a low-priority TODO

This whole issue is fairly high-priority. However, I'm personally currently working out details and quirks of Dream I/O streams and a Dream client, so I have to delay looking at sessions in detail again until I've taken care of those two things.
